test/utils/ktesting/ktesting.go

```diff
@@ -44,30 +41,14 @@ func SetDefaultVerbosity(v int) {
 	_ = f.Value.Set(fmt.Sprintf("%d", v))
 }
 
-// NewTestContext is a wrapper around ktesting.NewTestContext with settings
-// specific to Kubernetes.
-func NewTestContext(tl ktesting.TL) (klog.Logger, context.Context) {
-	config := ktesting.NewConfig(
-		ktesting.AnyToString(func(v interface{}) string {
-			return format.Object(v, 1)
-		}),
-		ktesting.VerbosityFlagName("v"),
-		ktesting.VModuleFlagName("vmodule"),
-	)
-
-	// Copy klog settings instead of making the ktesting logger
-	// configurable directly.
-	var fs flag.FlagSet
-	config.AddFlags(&fs)
-	for _, name := range []string{"v", "vmodule"} {
-		from := flag.CommandLine.Lookup(name)
-		to := fs.Lookup(name)
-		if err := to.Value.Set(from.Value.String()); err != nil {
-			panic(err)
-		}
-	}
-
-	logger := ktesting.NewLogger(tl, config)
-	ctx := klog.NewContext(context.Background(), logger)
-	return logger, ctx
+// NewTestContext is a drop-in replacement for ktesting.NewTestContext
+// which returns a more versatile context.
+//
+// The type of that context is still context.Context because replacing
+// it with TContext breaks tests which use `WithCancel`.
+//
+// TODO(pohly): change all of that code together with changing the return type.
```

Comment:
I have a follow-up commit that I intend to submit in a separate PR which does that across the code base:

```console
$ git log -p -n1 20a10317bbb1f81580f532d85aa94c46539233c2 | diffstat
 pkg/controller/disruption/disruption_test.go                               |   12 
 pkg/kubelet/config/config_test.go                                          |   60 ---
 test/integration/apiserver/apiserver_test.go                               |   39 --
 test/integration/apiserver/certreload/certreload_test.go                   |   20 -
 test/integration/apiserver/flowcontrol/concurrency_test.go                 |    9 
 test/integration/apiserver/flowcontrol/concurrency_util_test.go            |   11 
 test/integration/apiserver/max_json_patch_operations_test.go               |   11 
 test/integration/apiserver/max_request_body_bytes_test.go                  |   46 +-
 test/integration/apiserver/openapi/openapi_enum_test.go                    |    8 
 test/integration/apiserver/openapi/openapiv3_test.go                       |   14 
 test/integration/apiserver/podlogs/podlogs_test.go                         |   11 
 test/integration/apiserver/watchcache_test.go                              |   34 -
 test/integration/auth/accessreview_test.go                                 |   26 -
 test/integration/auth/auth_test.go                                         |   74 +---
 test/integration/auth/bootstraptoken_test.go                               |    8 
 test/integration/auth/dynamic_client_test.go                               |    9 
 test/integration/auth/rbac_test.go                                         |   48 +-
 test/integration/auth/selfsubjectreview_test.go                            |   22 -
 test/integration/auth/svcaccttoken_test.go                                 |   78 ++--
 test/integration/controlplane/synthetic_controlplane_test.go               |    7 
 test/integration/daemonset/daemonset_test.go                               |   18 -
 test/integration/defaulttolerationseconds/defaulttolerationseconds_test.go |   10 
 test/integration/disruption/disruption_test.go                             |   92 ++---
 test/integration/dualstack/dualstack_endpoints_test.go                     |   30 -
 test/integration/dualstack/dualstack_test.go                               |  180 ++++------
 test/integration/evictions/evictions_test.go                               |   88 ++--
 test/integration/examples/webhook_test.go                                  |   12 
 test/integration/framework/test_server.go                                  |    1 
 test/integration/network/services_test.go                                  |   60 +--
 test/integration/scheduler_perf/scheduler_test.go                          |    2 
 test/integration/serviceaccount/service_account_test.go                    |   56 +--
 test/integration/servicecidr/allocator_test.go                             |   20 -
 test/integration/util/util.go                                              |    9 
 test/utils/ktesting/ktesting.go                                            |    8 
 34 files changed, 447 insertions(+), 686 deletions(-)
```

Overall the code becomes shorter. The changes are like this one:
```patch
diff --git a/pkg/controller/disruption/disruption_test.go b/pkg/controller/disruption/disruption_test.go
index 3e9fb4e4fd3..223ac8dc3d8 100644
--- a/pkg/controller/disruption/disruption_test.go
+++ b/pkg/controller/disruption/disruption_test.go
@@ -1538,12 +1538,10 @@ func TestStalePodDisruption(t *testing.T) {
        }
        for name, tc := range cases {
                t.Run(name, func(t *testing.T) {
-                       _, ctx := ktesting.NewTestContext(t)
-                       ctx, cancel := context.WithCancel(ctx)
-                       defer cancel()
-                       dc, _ := newFakeDisruptionControllerWithTime(ctx, now)
-                       go dc.Run(ctx)
-                       if _, err := dc.coreClient.CoreV1().Pods(tc.pod.Namespace).Create(ctx, tc.pod, metav1.CreateOptions{}); err != nil {
+                       tCtx := ktesting.Init(t)
+                       dc, _ := newFakeDisruptionControllerWithTime(tCtx, now)
+                       go dc.Run(tCtx)
+                       if _, err := dc.coreClient.CoreV1().Pods(tc.pod.Namespace).Create(tCtx, tc.pod, metav1.CreateOptions{}); err != nil {
                                t.Fatalf("Failed to create pod: %v", err)
                        }
                        dc.clock.Sleep(tc.timePassed)
@@ -1552,7 +1550,7 @@ func TestStalePodDisruption(t *testing.T) {
                        }
                        diff := ""
                        if err := wait.Poll(100*time.Millisecond, wait.ForeverTestTimeout, func() (bool, error) {
-                               pod, err := dc.kubeClient.CoreV1().Pods(tc.pod.Namespace).Get(ctx, tc.pod.Name, metav1.GetOptions{})
+                               pod, err := dc.kubeClient.CoreV1().Pods(tc.pod.Namespace).Get(tCtx, tc.pod.Name, metav1.GetOptions{})
                                if err != nil {
                                        t.Fatalf("Failed getting updated pod: %v", err)
                                }
```

The decision to use `tCtx` instead of `ctx` become part of our "coding style". It would work to not rename the variable, then it's less clear when a `ctx` is a `context.Context` and when it is a `ktesting.TContext`. I prefer using `tCtx`.